components/core/src/clp_s/JsonParser.cpp

```diff
@@ -520,6 +534,265 @@
+            archive_node_type = NodeType::Boolean;
+            break;
+        case clp::ffi::SchemaTreeNode::Type::UnstructuredArray:
+            archive_node_type = NodeType::UnstructuredArray;
+            break;
+        case clp::ffi::SchemaTreeNode::Type::Str:
+            if (node_value && node_value->is<std::string>()) {
+                archive_node_type = NodeType::VarString;
+            } else {
+                archive_node_type = NodeType::ClpString;
+            }
+            break;
+        case clp::ffi::SchemaTreeNode::Type::Obj:
+            if (node_has_value) {
+                if (node_value->is_null()) {
+                    archive_node_type = NodeType::NullValue;
+                } else {
+                    archive_node_type = NodeType::Object;
+                }
+            } else {
+                archive_node_type = NodeType::Object;
+            }
+            break;
+        default:
+            break;
+    }
+    return archive_node_type;
+}
+
+auto JsonParser::get_archive_node_id(
+        std::unordered_map<int32_t, std::vector<std::pair<NodeType, int32_t>>>&
+                ir_node_to_archive_node_unordered_map,
+        int32_t ir_node_id,
+        NodeType archive_node_type,
+        clp::ffi::SchemaTree const& ir_tree
+) -> int {
+    auto unordered_map_location = ir_node_to_archive_node_unordered_map.find(ir_node_id);
+    if (ir_node_to_archive_node_unordered_map.end() != unordered_map_location) {
+        auto translation_vector = unordered_map_location->second;
+        for (int i = 0; i < translation_vector.size(); i++) {
+            if (translation_vector[i].first == archive_node_type) {
+                return translation_vector[i].second;
+            }
+        }
+    }
+
+    auto const& curr_node = ir_tree.get_node(ir_node_id);
+    int32_t parent_node_id{-1};
+    if (ir_node_id != curr_node.get_parent_id()) {
+        parent_node_id = get_archive_node_id(
+                ir_node_to_archive_node_unordered_map,
+                curr_node.get_parent_id(),
+                NodeType::Object,
+                ir_tree
+        );
+    }
+    auto validated_escaped_key
+            = clp::ffi::validate_and_escape_utf8_string(curr_node.get_key_name());
+    std::string node_key = "";
+    if (validated_escaped_key.has_value()) {
+        node_key = validated_escaped_key.value();
+    } else {
+        throw "Key is not UTF-8 compliant";
+    }
+    int curr_node_archive_id
+            = m_archive_writer->add_node(parent_node_id, archive_node_type, node_key);
+    auto p = std::make_pair(archive_node_type, curr_node_archive_id);
+    if (ir_node_to_archive_node_unordered_map.end() != unordered_map_location) {
+        unordered_map_location->second.push_back(p);
+    } else {
+        std::vector<std::pair<NodeType, int32_t>> v;
+        v.push_back(p);
+        ir_node_to_archive_node_unordered_map.emplace(ir_node_id, v);
+    }
+    return curr_node_archive_id;
+}
```

Comment:
1.) I mean to throw a message and not a std exception. 
2.) I named it this way on purpose because we changed the variable type
3.)I consider breaking into a smaller file at a later time.